dev/cs43l22/export/dev/cs43l22.hpp

```diff
@@ -0,0 +1,838 @@
+{
+    tx_cplt_cb();
+
+    ecl::err rc = I2c::init();
+    if (rc != ecl::err::ok) {
+        return rc;
+    }
+
+    rc = I2s::init();
+
+    if (is_ok(rc)) {
+        m_state |= inited;
+    }
+
+    return rc;
+}
+
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::power_up()
+{
+    ecl_assert(m_state & inited);
+
+    Rst_gpio::set();
+
+    uint8_t value {0};
+
+    // power up sequence according to RM
+    ecl::err rc = register_write(pwr_ctrl1, 0x01);
+    //TODO add error checking.
+    register_write(init_sequence_reg1, 0x99);
+    register_write(init_sequence_reg3, 0x80);
+
+    register_read(init_sequence_reg2, value);
+    register_write(init_sequence_reg2, value | 0x80);
+
+    register_read(init_sequence_reg2, value);
+    register_write(init_sequence_reg2, value & (~0x80));
+
+    register_write(init_sequence_reg1, 0x0);
+
+    // Turn ON headphone channel A and B
+    register_write(pwr_ctrl2, 0xAF);
+
+    // clock auto-detect
+    register_write(clk_ctrl, 0x81);
+
+    // Set Audio Word Length to 16 bit, Audio format - I2S.
+    register_write(if_ctrl1, 0x7);
+
+    // power up
+    register_write(pwr_ctrl1, 0x9e);
+
+    if (is_ok(rc)) {
+        m_state |= powered;
+    }
+
+    return ecl::err::ok;
+}
+
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::power_down()
+{
+    ecl_assert(m_state & inited);
+
+    ecl::err rc = register_write(pwr_ctrl1, 0x9f);
+    if (!is_ok(rc)) {
+        return rc;
+    }
+
+    Rst_gpio::reset();
+
+    m_state &= ~powered;
+
+    return rc;
+}
+
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::send_pcm_buffer(uint16_t *buffer, size_t count)
+{
+    ecl_assert(m_state & inited);
+
+    if (!buffer || !count) {
+        return ecl::err::inval;
+    }
+    if (m_state & tx_active) {
+        return ecl::err::busy;
+    }
+
+    m_state |= tx_active;
+
+    I2s::lock();
+
+    I2s::set_buffers((uint8_t *)buffer, nullptr, count * sizeof(*buffer));
+    ecl::err rc = I2s::xfer();
+
+    I2s::unlock();
+
+    m_state &= ~tx_active;
+
+    return rc;
+}
+
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::pcm_stream_start(uint16_t *buffer, size_t count, user_callback cb)
+{
+    ecl_assert(m_state & inited);
+
+    if (!buffer || !count) {
+        return ecl::err::inval;
+    }
+
+    if (m_state & tx_active) {
+        return ecl::err::busy;
+    }
+
+    m_state |= tx_active;
+
+    tx_cplt_cb() = cb;
+
+    I2s::lock();
+
+    I2s::platform_handle().enable_circular_mode();
+    I2s::set_buffers((uint8_t *)buffer, nullptr, count * sizeof(*buffer));
+    ecl::err rc = I2s::xfer(internal_tx_cplt_cb);
+
+    I2s::unlock();
+
+    return rc;
+}
+
+template < class I2c, class I2s, class Rst_gpio >
+ecl::err cs43l22< I2c, I2s, Rst_gpio >::pcm_stream_stop()
+{
+    ecl_assert(m_state & inited);
```

Comment:
Here. we need to ensure that `pcm_stream_start()` is actually called previously and stream wasn't stopped yet.